File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.py
import math
import heapq

def search(start, n, types, costs, answer):
    # Dijkstra
    intensities = [math.inf for i in range(n + 1)]
    intensities[start] = 0
    hq = [[0, start]] # [intensity, idx]
    
    while hq:
        now = heapq.heappop(hq)
        now_idx = now[1]
        now_intensity = now[0]
        
        # intensity가 가장 낮은 산봉오리에 도달하면 answer 확인 후 return
        if types[now_idx] == 2:
            if now_intensity < answer[1] or (now_intensity == answer[1] and now_idx < answer[0]):
                answer[0] = now_idx
                answer[1] = now_intensity
            return
        
        if now_intensity > intensities[now_idx]:
            continue
        for cost in costs[now_idx]:
            next_idx = cost[0]
            next_intensity = max(now_intensity, cost[1])
            
            # 출입구인 경우 계산하지 않음
            if types[next_idx] == 1:
                continue
            
            if next_intensity < intensities[next_idx]:
                intensities[next_idx] = next_intensity
                heapq.heappush(hq, [next_intensity, next_idx])

def solution(n, paths, gates, summits):
    answer = [-1, math.inf]
    
    types = [0 for i in range(n + 1)] # 0: 쉼터, 1: 출입구, 2: 산봉우리
    for gate in gates:
        types[gate] = 1
    for summit in summits:
        types[summit] = 2
    
    # 모든 cost 설정
    costs = [[] for i in range(n + 1)]
    for path in paths:
        costs[path[0]].append([path[1], path[2]])
        costs[path[1]].append([path[0], path[2]])
    
    # 각 gate에서 출발하여 확인
    for gate in gates:
        search(gate, n, types, costs, answer)
    
    return answer

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

struct cmp{
    bool operator()(vector<int> a, vector<int> b){
        if(a[1] == b[1]) return a[0] > b[0];
        return a[1] > b[1];
    } 
};

void search(int start, int n, vector<int>& types, vector<vector<vector<int>>>& costs, vector<int>& answer){
    // Dijkstra
    vector<int> intensities(n + 1, INT_MAX);
    intensities[start] = 0;
    priority_queue<vector<int>, vector<vector<int>>, cmp> pq; // {idx, intensity}
    pq.push({start, 0});
    
    while(!pq.empty()){
        int now = pq.top()[0];
        int now_intensity = pq.top()[1];
        pq.pop();
        
        if(types[now] == 2){
            if(now_intensity < answer[1] || (now_intensity == answer[1] && now < answer[0])){
                answer[0] = now;
                answer[1] = now_intensity;
            }
            return;
        }
        
        if(now_intensity > intensities[now]) continue;
        for(vector<int> cost : costs[now]){
            int next = cost[0];
            int next_intensity = max(now_intensity, cost[1]);
            if(types[next] == 1) continue;
            
            if(next_intensity < intensities[next]){
                intensities[next] = next_intensity;
                pq.push({next, next_intensity});
            }
        }
    }
}

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    vector<int> answer = {-1, INT_MAX};
    
    // 각 지점의 타입 설정
    vector<int> types(n + 1, 0); // 0: 쉼터, 1: 출입구, 2: 산봉우리
    for(int gate : gates) types[gate] = 1;
    for(int summit : summits) types[summit] = 2;
    
    // 모든 경로의 cost 설정
    vector<vector<vector<int>>> costs(n + 1, vector<vector<int>>());
    for(vector<int> path : paths){
        costs[path[0]].push_back({path[1], path[2]});
        costs[path[1]].push_back({path[0], path[2]});
    }
    
    // 모든 출발지점으로부터 계산
    for(int gate : gates){
        search(gate, n, types, costs, answer);
    }
    
    return answer;
}